### DISCUSSION TITLE
auto renewal

### DISCUSSION THREAD

**Discussion by leantalent** [GeorgeSchiro/GetCert2#16] [Q&A] [answered]
I use GetCert in the manual mode with several domains. Each time I have to edit the SAN list, GO then switch to the next domain.
How I can do this process with a script to make it automatic to say each month to renew a list of certificates ?

**Comment by GeorgeSchiro** [comment, marked as answer]
Look in [README.md](https://github.com/GeorgeSchiro/GetCert2/blob/master/README.md) for the section on "Command-Line Usage" Catalin. It shows you multiple ways to automate GC. One way is to simply maintain multiple versions of the GC profile file (ie. "GetCert2.exe.config").

To run GC with an alternative profile file, do this:

    GetCert2.exe -ini=NewProfile.txt

        or

    GetCert2.exe -ini=NewProfile.config

Each of the above examples assumes your alternative profile file is in the same folder as "GetCert2.exe", but of course you can place them anywhere and reference them with a full path, relative or absolute.

If you simply want an alternative SAN list (keeping everything else the same), you can do this:

    GetCert2.exe -SanList="-Domain='MyDomain.com' -Domain='www.MyDomain.com'"

Note: the two different types of delimiters above is intentional. The "outer delimiters" (double quotes in this case) must differ from the "inner delimiters" (single quotes in this case). The inner delimiters are only necessary if the domain name contains a space (impossible) or a hyphen (possible).

Of course, doing the above will still pop-up the green screen and require you to click "GO." If you really want to automate it, do this instead:

    GetCert2.exe -Auto -SanList="-Domain='MyDomain.com' -Domain='www.MyDomain.com'"

        or

    GetCert2.exe -Auto -ini=NewProfile.config

BTW, we will soon be releasing a newer version that is even more optimized for automation (especially high volumes - hundreds of domains). I would strongly recommend that you experiment now with the version you have and then replace it as soon as the new version becomes available. You do that simply by copying the new EXE over the old one Catalin.

**Comment by leantalent** [comment]
and when I have domain1.com and domain2.com and domain3.com how do I do it? Can I do this in a bat so that the next one start when the previous one ended?

GetCert2.exe -Auto -SanList="-Domain='Domain1.com' -Domain='www.MyDomain1.com'"
GetCert2.exe -Auto -SanList="-Domain='Domain2.com' -Domain='www.MyDomain2.com'"
GetCert2.exe -Auto -SanList="-Domain='Domain3.com' -Domain='www.MyDomain3.com'" 
...

**Reply by GeorgeSchiro** [reply]
Yes, exactly. I would recommend doing only one command-line first (whichever domain wasn't used during your previous manual run), just to be sure it works as expected.

I don't remember, but it's possible that version 2.5 doesn't support running several instances simultaneously. If that turns out to be the case, you may need to insert a "sleep" between each command-line. Here's what I mean:

    GetCert2.exe -Auto -SanList="-Domain='Domain1.com' -Domain='www.MyDomain1.com'"
    timeout 60
    GetCert2.exe -Auto -SanList="-Domain='Domain2.com' -Domain='www.MyDomain2.com'"
    timeout 60
    GetCert2.exe -Auto -SanList="-Domain='Domain3.com' -Domain='www.MyDomain3.com'"

Try it without the sleep. If that works, great.

**Reply by GeorgeSchiro** [reply]
I confirmed you don't need to sleep between command-lines. That said, if you intend to run your job unattended, then you will need one more switch:

    GetCert2.exe -Auto -NoPrompts -SanList="-Domain='Domain1.com' -Domain='www.MyDomain1.com'"
    GetCert2.exe -Auto -NoPrompts -SanList="-Domain='Domain2.com' -Domain='www.MyDomain2.com'"
    GetCert2.exe -Auto -NoPrompts -SanList="-Domain='Domain3.com' -Domain='www.MyDomain3.com'"

Also, if you plan to do several tests, you should do this (to avoid Let's Encrypt's rate limits):

    GetCert2.exe -Auto -DoStagingTests -SanList="-Domain='Domain1.com' -Domain='www.MyDomain1.com'"
    GetCert2.exe -Auto -DoStagingTests -SanList="-Domain='Domain2.com' -Domain='www.MyDomain2.com'"
    GetCert2.exe -Auto -DoStagingTests -SanList="-Domain='Domain3.com' -Domain='www.MyDomain3.com'"

**Comment by leantalent** [comment]
I did test with DoStagingTests as you suggested and I get 
2022-08-15 06:41:05:248 PM    "type": "urn:ietf:params:acme:error:malformed",
2022-08-15 06:41:05:248 PM    "detail": "upgrade your ACME client to support TLSv1.2 or better",
2022-08-15 06:41:05:248 PM    "status": 400

I did not used -NoPrompts but the .bat was executed without any prompts. I had to open the logs to see the error.

**Reply by GeorgeSchiro** [reply]
It's interesting that the ACME error doesn't happen when not running in staging mode, unless you are testing on a different server.

Anyway, try this (in your config file):

    -AcmeSystemWide='[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12'

**Reply by GeorgeSchiro** [reply]
In XML it will look like this:

    <add key="-AcmeSystemWide" value="[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12" />

**Comment by leantalent** [comment]
also I had to add  -CertificateDomainName since I saw that it was configured to the last one found in config.
GetCert2.exe -DoStagingTests -CertificateDomainName="domain1.com" -SanList="-Domain='domain1.com' -Domain='www.domain1.com'"

but the ACME error persists anyway

**Comment by GeorgeSchiro** [comment]
Yes, if the primary SAN list item doesn't match -CertificateDomainName, -CertificateDomainName will be updated to match.

If you will always run this stuff via your BAT file, no need to pass -CertificateDomainName on the command-line. The -CertificateDomainName value in the config will always show you which domain ran most recently.

**Comment by leantalent** [comment]
All good now. I had to put back in the config the -DoStagingTests="false" then only one domain was not updated GetCert2.exe -Auto -CertificateDomainName=i.domain.co" -SanList="-Domain='i.domanin.co'" but this I did it from console.
If I want to run this .bat in the background each month then I just have to set a task scheduler?

**Comment by GeorgeSchiro** [comment]
I should have been more precise.

If the primary SAN list item doesn't match -CertificateDomainName, -CertificateDomainName will be updated to match the first item. This is always the case unless -SanList only has one item. In that case, the mismatched -SanList is ignored and -CertificateDomainName takes precedence.

Yes, make sure all your command-lines include -NoPrompts and you can run the job via the task scheduler Catalin.
